File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
# from collections import defaultdict

def solution(genres, plays):
    answer = []
    total = {}
    seperate = {}
    for i in range(len(genres)):
        if genres[i] in total:
            total[genres[i]] += plays[i]
            seperate[genres[i]].append(i)
        else:
            total[genres[i]] = plays[i]
            seperate[genres[i]] = [i]
    total_list = []
    
    for key, value in total.items():
        total_list.append({key : value})
    total_list = sorted(total_list, key=lambda x: list(x.values())[0], reverse=True)
    
    print(total_list)
    
    for i in range(len(total_list)):
        best_g = list(total_list[i].keys())[0]
        best_i = sorted(seperate[best_g], key = lambda x:plays[x], reverse=True)
        if len(best_i) >= 2:
            answer.append(best_i[0])
            answer.append(best_i[1])
        else:
            answer.append(best_i[0])
    
    return answer

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.*;
import java.util.stream.Collectors;

class Solution {
    public int[] solution(String[] genres, int[] plays) {
        Map<String, Integer> l1 = new HashMap<>();
        Map<String, List<Integer>> l2 = new HashMap<>();
        for(int i = 0; i < genres.length; i++) {
            String g = genres[i];
            int p = plays[i];
            List<Integer> ex = l2.getOrDefault(g, new ArrayList<>());
            ex.add(i);
            
            l1.put(g, l1.getOrDefault(g, 0) + p);
            l2.put(g, ex);
        }
        l2.keySet().stream()
            .forEach(k -> l2.get(k).sort(Comparator.comparing((Integer i) -> plays[i]).reversed()));
        

        List<String> list = l1.keySet().stream()
            .sorted(Comparator.comparing(s -> l1.get(s)).reversed())
            .collect(Collectors.toList());
        
        
        List<Integer> answer = new ArrayList<>();
        for(int i = 0; i < list.size(); i++) {
            String g = list.get(i);
            List<Integer> nums = l2.get(g);
            for(int j = 0; j < Math.min(2, nums.size()); j++) {
                answer.add(nums.get(j));
            }
        }
        return answer.stream()
            .mapToInt(Integer::intValue)
            .toArray();
    }
}